### PR TITLE
Issues/4094 reader menu tags

### PR DESCRIPTION
**Pull request by aerych**
Refs #4094 
This is part two of the reader menu refresh. This PR reintroduces tag management. 
- Adds swipe to unfollow for tags you follow
- Adds a feature for following new tags.

To test:
- Ensure sure you can swipe to delete a tag you follow.
- Ensure default topics, lists, and the logged out tag list do not support swipe to delete.
- Ensure the confirmation prompt can be canceled. 
- Ensure that unfollowing a tag removes it from the list

Needs review: @kurzee would you be game to review?

### CODE REVIEW

**Comment by kurzee** [comment]
Crash while adding a tag for the first time, on iPhone 6 simulator.

```
2016-06-23 14:48:25.465 WordPress[66898:8071708] *** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndex:]: index 1 beyond bounds [0 .. 0]'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010ed9bd85 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x000000010e80fdeb objc_exception_throw + 48
	2   CoreFoundation                      0x000000010ec79804 -[__NSArrayM objectAtIndex:] + 212
	3   WordPress                           0x000000010764183d _TFC9WordPress19ReaderMenuViewModel21menuItemForTagAtIndexfSiGSqVS_14ReaderMenuItem_ + 989
	4   WordPress                           0x0000000107642a86 _TFC9WordPress19ReaderMenuViewModel19menuItemAtIndexPathfCSo11NSIndexPathGSqVS_14ReaderMenuItem_ + 998
	5   WordPress                           0x0000000107591b6f _TFC9WordPress24ReaderMenuViewController9tableViewfTCSo11UITableView21cellForRowAtIndexPathCSo11NSIndexPath_CSo15UITableViewCell + 143
	6   WordPress                           0x000000010759210f _TToFC9WordPress24ReaderMenuViewController9tableViewfTCSo11UITableView21cellForRowAtIndexPathCSo11NSIndexPath_CSo15UITableViewCell + 79
	7   UIKit                               0x000000010b71487c -[_UIFilteredDataSource tableView:cellForRowAtIndexPath:] + 98
	8   UIKit                               0x000000010b45b4f4 -[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:] + 766
	9   UIKit                               0x000000010b45b62c -[UITableView _createPreparedCellForGlobalRow:willDisplay:] + 74
	10  UIKit                               0x000000010b42fd4f -[UITableView _updateVisibleCellsNow:isRecursive:] + 2996
	11  UIKit                               0x000000010b464686 -[UITableView _performWithCachedTraitCollection:] + 92
	12  UIKit                               0x000000010b44b344 -[UITableView layoutSubviews] + 224
	13  UIKit                               0x000000010b3b8980 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 703
	14  QuartzCore                          0x000000010ad48c00 -[CALayer layoutSublayers] + 146
	15  QuartzCore                          0x000000010ad3d08e _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 366
	16  QuartzCore                          0x000000010ad3cf0c _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 24
	17  QuartzCore                          0x000000010ad313c9 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 277
	18  QuartzCore                          0x000000010ad5f086 _ZN2CA11Transaction6commitEv + 486
	19  QuartzCore                          0x000000010ad5f7f8 _ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv + 92
	20  CoreFoundation                      0x000000010ecc0c37 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	21  CoreFoundation                      0x000000010ecc0ba7 __CFRunLoopDoObservers + 391
	22  CoreFoundation                      0x000000010ecb67fb __CFRunLoopRun + 1147
	23  CoreFoundation                      0x000000010ecb60f8 CFRunLoopRunSpecific + 488
	24  GraphicsServices                    0x00000001116efad2 GSEventRunModal + 161
	25  UIKit                               0x000000010b2fdf09 UIApplicationMain + 171
	26  WordPress                           0x0000000107047a7e main + 62
	27  libdyld.dylib                       0x000000010fbab92d start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by kurzee** [comment]
After attempting to re-add the tag that caused the crash, which was `Apple`, the tag did not show in the list, and an error was received as:

```
Could not follow tag named Apple : Error Domain=WordPress.WordPressComRestApiError Code=6 "You are already subscribed to the specified tag." UserInfo={WordPressComRestApiErrorCodeKey=already_subscribed, WordPressComRestApiErrorMessageKey=You are already subscribed to the specified tag., NSLocalizedDescription=You are already subscribed to the specified tag.}
```

**Comment by kurzee** [comment]
After the crash above has occurred, even opening and dismissing the add tag view causes a crash, until it is deleted from the remote via the web:

```
2016-06-23 15:01:48.574 WordPress[72553:8100288] *** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndex:]: index 1 beyond bounds [0 .. 0]'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000107fe9d85 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x0000000107a5ddeb objc_exception_throw + 48
	2   CoreFoundation                      0x0000000107ec7804 -[__NSArrayM objectAtIndex:] + 212
	3   WordPress                           0x000000010088f82d _TFC9WordPress19ReaderMenuViewModel21menuItemForTagAtIndexfSiGSqVS_14ReaderMenuItem_ + 989
	4   WordPress                           0x0000000100890a76 _TFC9WordPress19ReaderMenuViewModel19menuItemAtIndexPathfCSo11NSIndexPathGSqVS_14ReaderMenuItem_ + 998
	5   WordPress                           0x00000001007dfb5f _TFC9WordPress24ReaderMenuViewController9tableViewfTCSo11UITableView21cellForRowAtIndexPathCSo11NSIndexPath_CSo15UITableViewCell + 143
	6   WordPress                           0x00000001007e00ff _TToFC9WordPress24ReaderMenuViewController9tableViewfTCSo11UITableView21cellForRowAtIndexPathCSo11NSIndexPath_CSo15UITableViewCell + 79
	7   UIKit                               0x000000010496287c -[_UIFilteredDataSource tableView:cellForRowAtIndexPath:] + 98
	8   UIKit                               0x00000001046a94f4 -[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:] + 766
	9   UIKit                               0x00000001046a962c -[UITableView _createPreparedCellForGlobalRow:willDisplay:] + 74
	10  UIKit                               0x000000010467dd4f -[UITableView _updateVisibleCellsNow:isRecursive:] + 2996
	11  UIKit                               0x00000001046b2686 -[UITableView _performWithCachedTraitCollection:] + 92
	12  UIKit                               0x0000000104699344 -[UITableView layoutSubviews] + 224
	13  UIKit                               0x0000000104606980 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 703
	14  QuartzCore                          0x0000000103f96c00 -[CALayer layoutSublayers] + 146
	15  QuartzCore                          0x0000000103f8b08e _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 366
	16  QuartzCore                          0x0000000103f8af0c _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 24
	17  QuartzCore                          0x0000000103f7f3c9 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 277
	18  QuartzCore                          0x0000000103fad086 _ZN2CA11Transaction6commitEv + 486
	19  QuartzCore                          0x0000000103fad7f8 _ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv + 92
	20  CoreFoundation                      0x0000000107f0ec37 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	21  CoreFoundation                      0x0000000107f0eba7 __CFRunLoopDoObservers + 391
	22  CoreFoundation                      0x0000000107f047fb __CFRunLoopRun + 1147
	23  CoreFoundation                      0x0000000107f040f8 CFRunLoopRunSpecific + 488
	24  GraphicsServices                    0x000000010a93dad2 GSEventRunModal + 161
	25  UIKit                               0x000000010454bf09 UIApplicationMain + 171
	26  WordPress                           0x0000000100295a3e main + 62
	27  libdyld.dylib                       0x0000000108df992d start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by kurzee** [comment]
It looks like the add new tag text field defaults to capitalization, maybe this should be lowercase? The web defaults to lowercase when adding a new tag.

**Comment by aerych** [comment]
Wicked crashes.  Digging into these.

**Comment by kurzee** [comment]
>Wicked crashes. Digging into these.
Crashes resolved. 👍

**Comment by kurzee** [comment]
Bonus: When adding a tag to an existing long list of tags, it would be awesome if the tableView would scroll down to show the newly added tag in the list.

**Comment by kurzee** [comment]
Nitpicky: You may have already looked at this, but would we want to still include the "Edit" bar button item? Swiping to unfollow is not necessarily obvious on my first impression.

**Comment by aerych** [comment]
> but would we want to still include the "Edit" bar button item

I was thinking no, but we certainly could.  @jancavan / @mattmiklic thoughts on this one?

**Comment by kurzee** [comment]
Seeing a view controller transition warning when attempting to add a tag that I've already added. Additionally the alert/error message itself does not present:

```
2016-06-23 15:35:35:982 WordPress[81928:8135056] followTagNamed:withSuccess:failure: error following topic by name: Error Domain=WordPress.WordPressComRestApiError Code=6 "You are already subscribed to the specified tag." UserInfo={WordPressComRestApiErrorCodeKey=already_subscribed, WordPressComRestApiErrorMessageKey=You are already subscribed to the specified tag., NSLocalizedDescription=You are already subscribed to the specified tag.}
2016-06-23 15:35:35:983 WordPress[81928:8135056] Could not follow tag named Apple : Error Domain=WordPress.WordPressComRestApiError Code=6 "You are already subscribed to the specified tag." UserInfo={WordPressComRestApiErrorCodeKey=already_subscribed, WordPressComRestApiErrorMessageKey=You are already subscribed to the specified tag., NSLocalizedDescription=You are already subscribed to the specified tag.}
2016-06-23 15:35:36.404 WordPress[81928:8135056] Warning: Attempt to present <UIAlertController: 0x7f8469d1e110> on <UINavigationController: 0x7f846b86f200> whose view is not in the window hierarchy!
```

Doesn't happen _every_ time, but perhaps a race condition between resigning the first responder and presenting the alert controller on the dismissing controller.

**Comment by kurzee** [comment]
Bonus: Upon first viewing the new menu, with tags, if none of the tags have already been loaded/cached, there is no loading indicator and there can be a few blank seconds of delay. (maybe an enhancement for later)

**Comment by aerych** [comment]
> Bonus: Upon first viewing the new menu, with tags, if none of the tags have already been loaded/cached, there is no loading indicator and there can be a few blank seconds of delay. (maybe an enhancement for later)

Totes!  Updating syncing behavior will come in pr numero cuatro.

File: WordPress/Classes/ViewRelated/Reader/ReaderMenuViewController.swift [hunk 1]
```diff
@@ -81,6 +85,100 @@ import WordPressShared
...
+    ///
+    func unfollowTagTopic(topic: ReaderTagTopic) {
+        let service = ReaderTopicService(managedObjectContext: ContextManager.sharedInstance().mainContext)
+        service.unfollowTag(topic, withSuccess: nil) { (error) in
+            DDLogSwift.logError("Could not unfollow topic \(topic), \(error)")
+
+            let title = NSLocalizedString("Could not Unfollow Tag", comment: "Title of a prompt informing the user there was a probem unsubscribing from a tag in the reader.")
+            let message = error.localizedDescription
+            let alert = UIAlertController(title: title, message: message, preferredStyle: .Alert)
+            alert.addCancelActionWithTitle(NSLocalizedString("OK", comment: "Button title. An acknowledge ment of the message displayed in a prompt."))
```
Comment (kurzee) [inline WordPress/Classes/ViewRelated/Reader/ReaderMenuViewController.swift]: Localized description typo: `An acknowledge ment`

File: WordPress/Classes/ViewRelated/Reader/ReaderMenuViewController.swift [hunk 2]
```diff
@@ -81,6 +85,100 @@ import WordPressShared
...
+    ///
+    func followTagNamed(tagName: String) {
+        let service = ReaderTopicService(managedObjectContext: ContextManager.sharedInstance().mainContext)
+        service.followTagNamed(tagName, withSuccess: nil) { (error) in
+            DDLogSwift.logError("Could not follow tag named \(tagName) : \(error)")
+
+            let title = NSLocalizedString("Could not Follow Tag", comment: "Title of a prompt informing the user there was a probem unsubscribing from a tag in the reader.")
+            let message = error.localizedDescription
+            let alert = UIAlertController(title: title, message: message, preferredStyle: .Alert)
+            alert.addCancelActionWithTitle(NSLocalizedString("OK", comment: "Button title. An acknowledge ment of the message displayed in a prompt."))
```
Comment (kurzee) [inline WordPress/Classes/ViewRelated/Reader/ReaderMenuViewController.swift]: Same thing on this one: ` An acknowledge ment`

File: WordPress/Classes/ViewRelated/Reader/ReaderMenuViewController.swift [hunk 3]
```diff
@@ -145,6 +256,63 @@ import WordPressShared
...
+        guard let menuItem = viewModel.menuItemAtIndexPath(indexPath) else {
+            return
+        }
+
+        WPStyleGuide.configureTableViewActionCell(cell)
+
+        let image = Gridicon.iconOfType(.AddOutline)
+        let imageView = UIImageView(image: image)
+        imageView.tintColor = WPStyleGuide.wordPressBlue()
+        cell.accessoryView = imageView
```
Comment (kurzee) [inline WordPress/Classes/ViewRelated/Reader/ReaderMenuViewController.swift]: Minor potential optimization here, the `accessoryView` may already be an `UIImageView` from a previous configure call. So we could technically save a view creation here and just set the image on the existing `UIImageView`.

File: WordPress/Classes/ViewRelated/Reader/ReaderMenuViewModel.swift [hunk 4]
```diff
@@ -403,7 +409,18 @@ enum ReaderDefaultMenuItemOrder: Int {
     /// - Returns: The requested menu item or nil.
     ///
     func menuItemForTagAtIndex(index: Int) -> ReaderMenuItem? {
-        guard let topic = tagsFetchedResultsController.objectAtIndexPath(NSIndexPath(forRow: index, inSection: 0)) as? ReaderAbstractTopic else {
+        var fetchedIndex = index
+        if ReaderHelpers.isLoggedIn() {
+            if fetchedIndex == 0 {
+                let title = NSLocalizedString("Add a Tag", comment: "Title. Let's the user know that they can use this fature to subscribe to new tags.")
```
Comment (kurzee) [inline WordPress/Classes/ViewRelated/Reader/ReaderMenuViewModel.swift]: Localization desc, `Let's the user know that they can use this fature`, maybe `Let's` > `Let` and `fature`.

**Comment by aerych** [comment]
> Doesn't happen every time, but perhaps a race condition between resigning the first responder and presenting the alert controller on the dismissing controller

The call to `presentFromRootViewController` was being silly and choosing a controller that was being dismissed.

**Comment by kurzee** [comment]
@aerych that's it from my end. Those other enhancements could come later as an enhancement. Let me know when you're ready for a final look.

**Comment by aerych** [comment]
Thanks @kurzee :)  I think I've addressed everything except the questions about scrolling to a newly added tag, and an edit button.  Let's give @mattmiklic / @jancavan a bit time to chime in.

**Comment by aerych** [comment]
@kurzee since there will be a couple more PRs following this one, and since its still feature flagged, let's call this one ready for a final round of code review and we'll address any feedback from @jancavan and @mattmiklic in a follow up.

**Comment by kurzee** [comment]
@aerych good to go :shipit:

**Comment by aerych** [comment]
Thanks @kurzee !!

**Comment by jancavan** [comment]
- I think we shouldn't capitalize "tag" here:

(1) "Add a tag" button in main tag list:
screenshot 2016-07-06 15 57 17

(2) Masterbar label:
screenshot 2016-07-06 16 22 24

(3) "Add tag" button:
screenshot 2016-07-06 16 22 27

(4) Invalid tag dialog box:
screenshot 2016-07-06 16 42 01

Should also just say:
(1) Could not follow tag (with F and T in small letters)
(2) "The tag requested does not seem to be valid." so we don't say "tag" 3 times

- We don't really introduce "Follow" in this particular view so it's confusing to randomly see "Unfollow" here. Let's just use "Remove" since we're also removing the tag from the list:
screenshot 2016-07-06 16 00 34

- When your list of tags has gotten really long and you add a new tag, it's not obvious that your new tag has been added. It would be nice if we could scroll it down to the position where the new tag is located. (When you add a new folder in iOS Notes app, it does this too).

- We should also just use the standard plus sign (without the circle) since we use that to "add" anything (e.g. add posts, pages, menu items) in Calypso:

screenshot 2016-07-06 16 38 13

**Comment by aerych** [comment]
Thanks for the review @jancavan! 

> I think we shouldn't capitalize "tag" here:

Apple's human interface guidelines calls for the use of title-case as opposed to sentence-case for each of these examples. Tho the current beta guidelines do say [sentence-case for Alerts](https://developer.apple.com/ios/human-interface-guidelines/ui-views/alerts/), I suspect its a typo since their example is in title-case.
Through out the rest of the app we use title-case for any button-like control, navbar title or alert title. I think it would look inconsistent to use sentence-case here. 

> (2) "The tag requested does not seem to be valid." so we don't say "tag" 3 times

The error message is actually returned from the API.  We can [open an issue](https://github.com/Automattic/wp-calypso/issues/6609) to change it there.

> We don't really introduce "Follow" in this particular view so it's confusing to randomly see "Unfollow" here. Let's just use "Remove" since we're also removing the tag from the list

👍 

> When your list of tags has gotten really long and you add a new tag, it's not obvious that your new tag has been added. It would be nice if we could scroll it down to the position where the new tag is located

👍 

> We should also just use the standard plus sign 

👍

**Comment by jancavan** [comment]
@aerych Got it. Title case sounds good as look as there's uniformity everywhere 👍 

On the invalid dialog tag box, the word tag is also all in caps.

**Comment by aerych** [comment]
> On the invalid dialog tag box, the word tag is also all in caps.

@tootallnate took care of this one on the [api side of things](https://github.com/Automattic/wp-calypso/issues/6609#issuecomment-231181982).

**Pull request merged**